File: server/camcops_server/tasks/photo.py
#!/usr/bin/env python

"""
camcops_server/tasks/photo.py

===============================================================================

    Copyright (C) 2012-2019 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================

"""

# import logging
from typing import List, Optional

# from cardinal_pythonlib.logs import BraceStyleAdapter
import cardinal_pythonlib.rnc_web as ws
from sqlalchemy.sql.schema import Column
from sqlalchemy.sql.sqltypes import Integer, UnicodeText

from camcops_server.cc_modules.cc_blob import (
    Blob,
    blob_relationship,
    get_blob_img_html,
)
from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_ctvinfo import CTV_INCOMPLETE, CtvInfo
from camcops_server.cc_modules.cc_db import (
    ancillary_relationship,
    GenericTabletRecordMixin,
)
from camcops_server.cc_modules.cc_html import answer, tr_qa
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_snomed import SnomedExpression, SnomedLookup
from camcops_server.cc_modules.cc_sqla_coltypes import CamcopsColumn
from camcops_server.cc_modules.cc_sqlalchemy import Base
from camcops_server.cc_modules.cc_task import (
    Task,
    TaskHasClinicianMixin,
    TaskHasPatientMixin,
)

# log = BraceStyleAdapter(logging.getLogger(__name__))


# =============================================================================
# Photo
# =============================================================================

class Photo(TaskHasClinicianMixin, TaskHasPatientMixin, Task):
    """
    Server implementation of the Photo task.
    """
    __tablename__ = "photo"
    shortname = "Photo"

    description = Column(
        "description", UnicodeText,
        comment="Description of the photograph"
    )
    photo_blobid = CamcopsColumn(
        "photo_blobid", Integer,
        is_blob_id_field=True, blob_relationship_attr_name="photo",
        comment="ID of the BLOB (foreign key to blobs.id, given "
                "matching device and current/frozen record status)"
    )
    # IGNORED. REMOVE WHEN ALL PRE-2.0.0 TABLETS GONE:
    rotation = Column(  # DEFUNCT as of v2.0.0
        "rotation", Integer,
        comment="Rotation (clockwise, in degrees) to be applied for viewing"
    )

    photo = blob_relationship("Photo", "photo_blobid")  # type: Optional[Blob]

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Photograph")

    def is_complete(self) -> bool:
        return self.photo_blobid is not None

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        if not self.description:
            return []
        return [CtvInfo(content=self.description)]

    def get_task_html(self, req: CamcopsRequest) -> str:
        # noinspection PyTypeChecker
        return """
            <table class="{CssClass.TASKDETAIL}">
                <tr class="{CssClass.SUBHEADING}"><td>Description</td></tr>
                <tr><td>{description}</td></tr>
                <tr class="{CssClass.SUBHEADING}"><td>Photo</td></tr>
                <tr><td>{photo}</td></tr>
            </table>
        """.format(
            CssClass=CssClass,
            description=answer(
                ws.webify(self.description), default="(No description)",
                default_for_blank_strings=True
            ),
            # ... xhtml2pdf crashes if the contents are empty...
            photo=get_blob_img_html(self.photo)
        )

    def get_snomed_codes(self, req: CamcopsRequest) -> List[SnomedExpression]:
        if not self.is_complete():
            return []
        return [
            SnomedExpression(req.snomed(SnomedLookup.PHOTOGRAPH_PROCEDURE)),
            SnomedExpression(req.snomed(SnomedLookup.PHOTOGRAPH_PHYSICAL_OBJECT)),  # noqa
        ]


# =============================================================================
# PhotoSequence
# =============================================================================

class PhotoSequenceSinglePhoto(GenericTabletRecordMixin, Base):
    __tablename__ = "photosequence_photos"

    photosequence_id = Column(
        "photosequence_id", Integer, nullable=False,
        comment="Tablet FK to photosequence"
    )
    seqnum = Column(
        "seqnum", Integer, nullable=False,
        comment="Sequence number of this photo "
                "(consistently 1-based as of 2018-12-01)"
    )
    description = Column(
        "description", UnicodeText,
        comment="Description of the photograph"
    )
    photo_blobid = CamcopsColumn(
        "photo_blobid", Integer,
        is_blob_id_field=True, blob_relationship_attr_name="photo",
        comment="ID of the BLOB (foreign key to blobs.id, given "
                "matching device and current/frozen record status)"
    )
    # IGNORED. REMOVE WHEN ALL PRE-2.0.0 TABLETS GONE:
    rotation = Column(  # DEFUNCT as of v2.0.0
        "rotation", Integer,
        comment="(DEFUNCT COLUMN) "
                "Rotation (clockwise, in degrees) to be applied for viewing"
    )

    photo = blob_relationship("PhotoSequenceSinglePhoto", "photo_blobid")

    def get_html_table_rows(self) -> str:
        # noinspection PyTypeChecker
        return """
            <tr class="{CssClass.SUBHEADING}">
                <td>Photo {num}: <b>{description}</b></td>
            </tr>
            <tr><td>{photo}</td></tr>
        """.format(
            CssClass=CssClass,
            num=self.seqnum,
            description=ws.webify(self.description),
            photo=get_blob_img_html(self.photo)
        )


class PhotoSequence(TaskHasClinicianMixin, TaskHasPatientMixin, Task):
    """
    Server implementation of the PhotoSequence task.
    """
    __tablename__ = "photosequence"
    shortname = "PhotoSequence"

    sequence_description = Column(
        "sequence_description", UnicodeText,
        comment="Description of the sequence of photographs"
    )

    photos = ancillary_relationship(
        parent_class_name="PhotoSequence",
        ancillary_class_name="PhotoSequenceSinglePhoto",
        ancillary_fk_to_parent_attr_name="photosequence_id",
        ancillary_order_by_attr_name="seqnum"
    )  # type: List[PhotoSequenceSinglePhoto]

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Photograph sequence")

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        infolist = [CtvInfo(content=self.sequence_description)]
        for p in self.photos:
            infolist.append(CtvInfo(content=p.description))
        return infolist

    def get_num_photos(self) -> int:
        return len(self.photos) > 0

    def is_complete(self) -> bool:
        # If you're wondering why this is being called unexpectedly: it may be
        # because this task is being displayed in the task list, at which point
        # we colour it by its complete-or-not status.
        return bool(self.sequence_description) and self.get_num_photos() > 0

    def get_task_html(self, req: CamcopsRequest) -> str:
        html = f"""
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {self.get_is_complete_tr(req)}
                    {tr_qa("Number of photos", self.get_num_photos())}
                    {tr_qa("Description", self.sequence_description)}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
        """
        for p in self.photos:
            html += p.get_html_table_rows()
        html += """
            </table>
        """
        return html

    def get_snomed_codes(self, req: CamcopsRequest) -> List[SnomedExpression]:
        if not self.is_complete():
            return []
        return [
            SnomedExpression(req.snomed(SnomedLookup.PHOTOGRAPH_PROCEDURE)),
            SnomedExpression(req.snomed(SnomedLookup.PHOTOGRAPH_PHYSICAL_OBJECT)),  # noqa
        ]

File: tablet_qt/tasks/photo.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

// #define DEBUG_IS_COMPLETE

#include "photo.h"

#include "common/textconst.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/quphoto.h"
#include "questionnairelib/qutext.h"
#include "questionnairelib/qutextedit.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"

const QString Photo::PHOTO_TABLENAME("photo");

const QString DESCRIPTION("description");
const QString PHOTO_BLOBID("photo_blobid");

// const QString ROTATION("rotation");  // DEFUNCT in v2


void initializePhoto(TaskFactory& factory)
{
    static TaskRegistrar<Photo> registered(factory);
}

Photo::Photo(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, PHOTO_TABLENAME, false, true, false)  // ... anon, clin, resp
{
    addField(DESCRIPTION, QMetaType::fromType<QString>());
    addField(PHOTO_BLOBID, QMetaType::fromType<int>());  // FK to BLOB table
    // addField(ROTATION, QMetaType::fromType<int>());  // DEFUNCT in v2

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Photo::shortname() const
{
    return "Photo";
}

QString Photo::longname() const
{
    return tr("Photograph");
}

QString Photo::description() const
{
    return tr("Photograph with accompanying detail.");
}

QString Photo::infoFilenameStem() const
{
    return "clinical";
}

// ============================================================================
// Instance info
// ============================================================================

bool Photo::isComplete() const
{
#ifdef DEBUG_IS_COMPLETE
    qDebug() << "valueIsNullOrEmpty(DESCRIPTION)"
             << valueIsNullOrEmpty(DESCRIPTION);
    qDebug() << "valueIsNull(PHOTO_BLOBID)" << valueIsNull(PHOTO_BLOBID);
#endif
    return !valueIsNullOrEmpty(DESCRIPTION) && !valueIsNull(PHOTO_BLOBID);
}

QStringList Photo::summary() const
{
    return QStringList{valueString(DESCRIPTION)};
}

QStringList Photo::detail() const
{
    return completenessInfo() + summary();
}

OpenableWidget* Photo::editor(const bool read_only)
{
    QuPagePtr page(
        (new QuPage{
             new QuText(tr("1. Ensure consent is documented, if applicable.\n"
                           "2. Take a photograph.\n"
                           "3. Enter a description.")),
             new QuText(TextConst::description()),
             new QuTextEdit(fieldRef(DESCRIPTION)),
             new QuPhoto(blobFieldRef(PHOTO_BLOBID, false)),
         })
            ->setTitle(tr("Clinical photograph"))
    );

    auto questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Clinician);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}
